### ISSUE TITLE
Preprocess several files

### ISSUE THREAD

**Issue by hendiol** [evanplaice/pypreprocessor#6]
How to preprocess several files after each other in one script ?

I had to make a litte change in the: File: __ init __.py ; Function: post_process(self) [deleted the sys.exit(0) statement if run==False ] and adding a function:
Original:
```
        # resolve postprocess stage depending on the mode
        if self.run == False:
            sys.exit(0)
        else:
```

Changed:
```
        # resolve postprocess stage depending on the mode
        if self.run:
```
Added Function:
```
    # reseting internal things to parse a second file
    def reset_internal(self):
        self.__linenum = 0
        self.__excludeblock = False
        self.__ifblock = False
        self.__ifcondition = ''
        self.__ifconditions = []
        self.__evalsquelch = True
        self.__outputBuffer = ''
```

now I can do things like this: 
pypreprocessor.input = 'Algorithm.py'
pypreprocessor.output = 'Algorithm_out.py'
pypreprocessor.parse()

pypreprocessor.input = 'Algo_2.py'
pypreprocessor.output = 'Algo_2_out.py'
pypreprocessor.parse()

Process several file with the same #ifdef statements in one file is there another oportunity to do so ?

**Comment by evanplaice** [comment]
There's probably a better way to do this. For instance make the input and output a list.

**Comment by hendiol** [comment]
Yes i can do this by lists, but this does not change the original problem that my code stops after preprocessing the first file (Without the changes mentioned above in your package (deleting the sys.exit(0) statement))

Here is a minimal example (works fine with my changes, not with original package, and I am asking for  a better way to do this ?):

Main:
`from pypreprocessor import pypreprocessor

options=[['rain',False],['debug',False]]

print('Starting prepro')

for opt in options:
    if opt[1]:
        pypreprocessor.defines.append(opt[0])
#pypreprocessor.removeMeta = True

codes=['Algorithm', 'Algorithm_2']
for c in codes:
    pypreprocessor.input = c+'.py'
    pypreprocessor.output = c+'_out.py'
    pypreprocessor.parse()
    pypreprocessor.reset_internal()
print('Finished pre pro')



print('Now doing')
import Algorithm_out
Algorithm_out.do()


print('Now doing 2')
import Algorithm_2_out
Pferd=Algorithm_2_out.funny_horse(9)
Pferd.say_age()
Pferd.say_feeling()

`
Algorithm.py
`def do():
#ifdef debug
    print('The source is in debug mode')
#else
    print('The source is not in debug mode')
#endif
   
#ifdef rain
    print('It is raining')
#else
    print('Dry')
#endif`

Algorithm_2.py
`class funny_horse():
    def __init__(self,age):
        self.alter=age
    def say_age(self):
        print('I´m the funny horse and may age is ',self.alter)
    def say_feeling(self):
#ifdef rain
        print('I´m sad, because it is raining here and i am very wet')
#else
        print('I´m happy, the sun is shining')
#endif`

This is the output with my changes:
Console:   python main.py
Starting prepro
Finished pre pro
Now doing
The source is not in debug mode
Dry
Now doing 2
I´m the funny horse and may age is  9
I´m happy, the sun is shining

and without:
Console: python main.py
Starting prepro

So without the changes, the code terminates while preprocessing.

**Comment by evanplaice** [comment]
`self.run is False` is telling the preprocessor that the output files will be produced but not run. This works more like a transpiler (ex sass, less, babel), where the files are preprocessed on-the-fly and the output is saved as new files.

*Note: This can be a great approach if you setup a simple file watcher that automatically preprocesses the files when it detects that the source has been changed.*

`self.run is True' is telling the preprocessor to run the postprocessed source of the file it's defined in. It preprocesses the source on-the-fly, imports, and launches it all at once. This works more like C where the preprocessor directives are supported natively, except python doesn't allow preprocessor code so this requires a bit of magic to work.

*Note: This approach is ideal if you just need some simple branch logic in a single source file. That doesn't need to be imported by something else.*

The 2nd approach won't work because the python's internal lexer/parser processes all import statements on the first pass. In this case, you're trying to import source files that don't exist yet (ie the preprocessor doesn't run until after python's internal lexer/parser). There is a way to modify the internal imports list after python's internal lexer/parser has run, that's the 'magic' that makes on-the-fly mode work. To see it in action, check out the `override_imports` method.

In short, if you want to load files in the same source that is responsible for preprocessing them, you have to use an alternative to the `import` statement. Things that seem very straight-forward and simple don't necessarily work as you'd expect them to when you're trying to load code that's being dynamically generated.

**Comment by hendiol** [comment]
I am using pypreprocessor as transpiller (called in another python file (main)), saving the files, and afterwards importing this newly created files in main and then using the preprocessed and afterwards included file. With the statement: `self.run is False` and then `sys.exit(0)` The main file terminates after the first file is preprocessed. Therefore I forked your project and added an option to the parse function, handed over to post_process and dissabling the `sys.exit(0)`. This works fine for me, I don't see the problems mentioned above.
Also I added the possibility to have nested `#ifdef` statements

**Comment by evanplaice** [comment]
Makes sense. Using it to externally transpile and later import fits the intended usage.

If you want to cut down the size of output source you could set `removeMeta` to True. That will remove all directives and unused code from the output. Although it's not really necessary; once the output source is run once the .pyc will contain only bytecode.

AFAIK the `sys.exit(0)` is unnecessary and worth removing if it's causing issues. Submit that as a PR and I'll merge it.

For the nested `#ifdef` how does that look in practice? I'm good with adding that capability if it's useful. The usage just needs to be documented in the README.

If you'd like to make further improvements, let me know and I'll grant push access. The only thing I'm a stickler about is logically separating changes into different commits and documenting each commit with a relevant message.

**Comment by hendiol** [comment]
Thanks for the answers to evanplaice.

OK, fine.

Yes I know the function, for debugging reasons i often left `removeMeta = False` (so that the line numbers are consistend)

OK, I added a varaible mode with wich the user can swith between: 'Run' (is the same as your` run = True`), 'PP' (preprocess only is the same as before `run = False`, PPCont (preprocess and continue: preprocess file and reset internal variables)

The nested `#ifdef`  works with a list, you can see it in my project. I will modify my Readme and the examples and afterwards do a PR. (Some test on my version are done but if you have some cases you can test them)

Oh, i hope my commits into my PR are not to weird. At the moment I have no further ideas.
